cpufeatures/src/aarch64.rs

```diff
@@ -0,0 +1,91 @@
+/// APIs for feature detection.
+
+/// Create module with CPU feature detection code.
+#[macro_export]
+macro_rules! new {
+    ($mod_name:ident, $($tf:tt),+ $(,)? ) => {
+        mod $mod_name {
+            use core::sync::atomic::{AtomicU8, Ordering::Relaxed};
+
+            const UNINIT: u8 = u8::max_value();
+            static STORAGE: AtomicU8 = AtomicU8::new(UNINIT);
+
+            /// Initialization token
+            #[derive(Copy, Clone, Debug)]
+            pub struct InitToken(());
+
+            impl InitToken {
+                /// Get initialized value
+                #[inline(always)]
+                pub fn get(&self) -> bool {
+                    // CPUID is not available on SGX targets
+                    #[cfg(not(all($(target_feature=$tf, )*)))]
+                    let res = STORAGE.load(Relaxed) == 1;
+                    #[cfg(all($(target_feature=$tf, )*))]
+                    let res = true;
+                    res
+                }
+            }
+
+            /// Initialize underlying storage if needed and get
+            /// stored value and initialization token.
+            #[inline]
+            pub fn init_get() -> (InitToken, bool) {
+                #[cfg(not(all($(target_feature=$tf, )*)))]
+                let res = {
+                    // Relaxed ordering is fine, as we only have a single atomic variable.
+                    let val = STORAGE.load(Relaxed);
+                    if val == UNINIT {
+                        #[allow(unused_variables)]
+                        let hwcaps = unsafe { libc::getauxval(libc::AT_HWCAP) };
+                        let res = $(cpufeatures::check!(hwcaps, $tf) & )+ true;
+                        STORAGE.store(res as u8, Relaxed);
+                        res
+                    } else {
+                        val == 1
+                    }
+                };
+                #[cfg(all($(target_feature=$tf, )*))]
+                let res = true;
+
+                (InitToken(()), res)
+            }
+
+            /// Initialize underlying storage if needed and get
+            /// initialization token.
+            #[inline]
+            pub fn init() -> InitToken {
+                init_get().0
+            }
+
+            /// Initialize underlying storage if needed and get
+            /// stored value.
+            #[inline]
+            pub fn get() -> bool {
+                init_get().1
+            }
+        }
+    };
+}
+
+macro_rules! expand_check_macro {
+    ($(($name:tt, $hwcap:expr)),* $(,)?) => {
+        #[macro_export]
+        #[doc(hidden)]
+        macro_rules! check {
+            $(
+                ($hwcaps:expr, $name) => { (($hwcaps & libc::$name) != 0) };
+            )*
+        }
+    };
+}
+
+expand_check_macro! {
+    ("aes",  HWCAP_AES),  // Enable AES support.
+    ("sha2", HWCAP_SHA2), // Enable SHA1 and SHA256 support.
+    ("sha3", HWCAP_SHA3), // Enable SHA512 and SHA3 support.
```

Comment:
Note: the comments here are taken directly from the output of:

```
$ rustc --target=aarch64-unknown-linux-gnu --print target-features
```

I'm a bit confused because the target features aren't isomorphic to either the Linux ARM64 ELF hwcaps or the macOS sysctl flags.

Per the Linux ARM64 ELF hwcaps docs:

https://www.kernel.org/doc/html/latest/arm64/elf_hwcaps.html

> HWCAP_SHA1: Functionality implied by ID_AA64ISAR0_EL1.SHA1 == 0b0001.
> HWCAP_SHA2: Functionality implied by ID_AA64ISAR0_EL1.SHA2 == 0b0001.
> HWCAP_SHA3: Functionality implied by ID_AA64ISAR0_EL1.SHA3 == 0b0001.
> HWCAP_SHA512: Functionality implied by ID_AA64ISAR0_EL1.SHA2 == 0b0010.

That would suggest that the SHA-1 and SHA-3 hwcaps are independent of the others, and that SHA-2 support implies SHA-512 support.

macOS/M4 likewise has separate `sysctl(3)` flags for SHA-512 and SHA-3:

```
hw.optional.armv8_2_sha512: 1
hw.optional.armv8_2_sha3: 1
```